pkg/association/association.go

```diff
@@ -16,195 +16,280 @@ import (
 	retryInterval := smf_context.GetSelf().AssocFailRetryInterval
 	for {
 		timer := time.After(retryInterval)
-		err := setupPfcpAssociation(upf, upfStr)
+		err := setupPfcpAssociation(upf)
 		if err == nil {
+			// success
 			return
 		}
-		logger.MainLog.Warnf("Failed to setup an association with UPF%s, error:%+v", upfStr, err)
+		logger.PfcpLog.Warnf("Failed to setup an association with UPF[%s], error:%+v", upf.GetNodeIDString(), err)
 		now := time.Now()
-		logger.MainLog.Debugf("now %+v, alertTime %+v", now, alertTime)
+		logger.PfcpLog.Debugf("now %+v, alertTime %+v", now, alertTime)
 		if now.After(alertTime.Add(alertInterval)) {
-			logger.MainLog.Errorf("ALERT for UPF%s", upfStr)
+			logger.PfcpLog.Errorf("ALERT for UPF[%s]", upf.GetNodeIDString())
 			alertTime = now
 		}
-		logger.MainLog.Debugf("Wait %+v (or less) until next retry attempt", retryInterval)
+		logger.PfcpLog.Debugf("Wait %+v (or less) until next retry attempt", retryInterval)
 		select {
 		case <-ctx.Done():
-			logger.MainLog.Infof("Canceled association request to UPF%s", upfStr)
+			logger.PfcpLog.Infof("Canceled smf context, stop association request to UPF[%s]", upf.GetNodeIDString())
 			return
-		case <-upf.Ctx.Done():
-			logger.MainLog.Infof("Canceled association request to this UPF%s only", upfStr)
+		case <-upf.Association.Done():
+			logger.MainLog.Infof("Canceled association request to this UPF%s only", upf.GetNodeIDString())
 			return
 		case <-timer:
 			continue
+		default:
 		}
 	}
 }
 
-func setupPfcpAssociation(upf *smf_context.UPF, upfStr string) error {
-	logger.MainLog.Infof("Sending PFCP Association Request to UPF%s", upfStr)
+func setupPfcpAssociation(upf *smf_context.UPF) error {
+	logger.PfcpLog.Infof("Send an Association Setup Request to UPF[%s]", upf.GetNodeIDString())
 
-	resMsg, err := message.SendPfcpAssociationSetupRequest(upf.NodeID)
+	resMsg, err := message.SendPfcpAssociationSetupRequest(upf.PFCPAddr())
 	if err != nil {
 		return err
 	}
 
 	rsp := resMsg.PfcpMessage.Body.(pfcp.PFCPAssociationSetupResponse)
 
 	if rsp.Cause == nil || rsp.Cause.CauseValue != pfcpType.CauseRequestAccepted {
-		return fmt.Errorf("received PFCP Association Setup Not Accepted Response from UPF%s", upfStr)
+		return fmt.Errorf("received PFCP Association Setup Not Accepted Response from UPF[%s]", upf.GetNodeIDString())
 	}
 
-	nodeID := rsp.NodeID
-	if nodeID == nil {
+	if rsp.NodeID == nil {
 		return fmt.Errorf("pfcp association needs NodeID")
 	}
 
-	logger.MainLog.Infof("Received PFCP Association Setup Accepted Response from UPF%s", upfStr)
+	logger.PfcpLog.Infof("Received PFCP Association Setup Accepted Response from UPF[%s]", upf.GetNodeIDString())
 
 	upf.UPFStatus = smf_context.AssociatedSetUpSuccess
 
 	if rsp.UserPlaneIPResourceInformation != nil {
 		upf.UPIPInfo = *rsp.UserPlaneIPResourceInformation
+		logger.MainLog.Infof("UPF(%s)[%s] setup association", upf.GetNodeIDString(), upf.UPIPInfo.NetworkInstance.NetworkInstance)
```

Comment:
Hi @LaumiH 
After the refactoring, SMF will try to restore the session if the associated UPF hasn't responded to the heartbeat request.
However, UDP packets may be affected by network burst/congestion. In this case, UPF may have a short time disconnected from SMF, we should not reset its session context.

Thus,  I think we can consider making UPF support to send association setup (with timestamp IE) to help SMF identify the defeated cause. Do you have any comments on this?
Thanks!

---
@andy89923 FYI.

Reply:
Hi @ianchen0119, I am not sure I completely get your idea. Let me describe what I understood: 

Currently, a UPF resets its session contexts entirely after association to the SMF, regardless of previously existing sessions. But when a heartbeat reply is only missing because of the network and not because of a UPF crash, all sessions need to be re-established again.

I think in such a case it would make sense to not reset the session contexts on the UPF and avoid having to restore all sessions through the session establishment procedure.

The timeline would look like this:
1. heartbeat miss due to network problems (heartbeat timeout exceeded), and not due to UPF crash
2. SMF detects UPF failure, but retains all PFCPSessionContexts related to this UPF for later session restoration
3. SMF send (re-) association request to the UPF
4. UPF realizes (through this association request) that the SMF falsely detected a UPF failure
5. UPF does not delete its session context
6. UPF informs SMF about existing session context through timestamp IE (timestamp of last modified session context?)
7. SMF checks session context sync: if there have been newer PFCPSessionContexts (> timestamp), only send those to the UPF, otherwise skip session restoration procedure

Does this match your idea?

Reply:
Hi @LaumiH 
In my opinion,
It would be great if UPF supports initiating an Association Request.
In my scenario,
UPF will:
1. receive an association request from SMF and send an association response with its recovery timestamp.
2. send an association request to SMF with its recovery timestamp.

during the initialization procedure.

At the same time, SMF can treat the association request and response sent from UPF as one association procedure based on UPF's recovery timestamp.
Once UPF restarted, it will send an association request to SMF with a new timestamp immediately, then SMF initiates the session restoration procedure because UPF just restarted.

As for the heartbeat timeout exceeded, SMF should retain the session context(s) and try to do re-association.
(Further behaviors are the same with the initialization procedure)

Reply:
Hi @ianchen0119, so the session restoration procedure will be triggered by the UPF instead of the SMF?

Reply:
Hi @LaumiH 

Sorry for late reply.
I'm very busy this week..
I'll discuss with team members and get back to you soon🙏

Reply:
Hello @LaumiH Thanks for your PR.

I would like to provide some suggestions regarding UPF Restoration.
According to Spec **TS 23.527 v15 4.2** mentioned **N4 Failure and Restart Detection**,
SMF shall utilize the PFCP Heartbeat Request and Heartbeat Response messages with PFCP messages containing the Recovery Time Stamp to detect whether a peer PFCP failed or restarted.

Currently, free5GC stores the `RecoveryTimeStamp` value in the `smfcontext.UPF` data structure and checks if the UPF has restarted in the function `doPfcpHeartbeat` (`upf.RecoveryTimeStamp.Before(rsp.RecoveryTimeStamp.RecoveryTimeStamp)`).
However, according to the Spec,
when the SMF detects a UPF restart, the SMF should reinitiate a pfcpAssociationSetupRequest, which is not yet implemented in free5GC.
In your PR, you mentioned that you implemented this part, but it seems it is not based on the **RecoveryTimeStamp.** Please help to confirm it.

Additionally, according to Spec **TS23.527 v15 4.3.2**,
SMF may start restoring PFCP sessions in the UPF immediately after the re-establishment of a PFCP association between the SMF and the UPF. This means that the SMF needs to store the PFCP Session for each individual UPF (you can refer to the `PFCPContext` value, which is stored in the `SMContext` data structure)
If you have added this part, please let us know.

By the way, recently free5GC has undergone refactor updates for various Network Functions. If possible, please make your modifications based on the latest commits. Thank you.

Reply:
Hi @ming-hsien , I want to address your points:
1. I am not making use of the RecoveryTimeStamp yet in my refactored code. However, the SMF will send the pfcpAssociationSetupRequest again and reassociate as soon as the UPF answers.
2. I am using the same variable in the SMContext data structure to store the PFCP session contexts individually for each UPF. In addition to that (to speed up some logic), each UPF object stores a map with a pointer to the same PFCP session context variable. I use this to restore all previous session contexts on a UPF that has been re-associated after reboot. Without such a data structure, restoring sessions on a UPF would be impossible in my code :)
3. As soon as I have the time, I'll rebase my code on the current SMF code and especially check if the communication to the other NFs (AMF, PCF etc.) changed.

Reply:
Hello @LaumiH , your understanding is correct. 
Additionally, according to what you described in your second point, we will review your PR again after you update it to the new version of free5GC. Thank you =)